JSON.GET with filter fails on NoClassDefFound: net/minidev/json/parser/ParseException

### Describe the bug

Using filters in JSON.GET results either in incorrect response or `NoClassDefFound: net/minidev/json/parser/ParseException'.

**Test:** `tests/pytests/test.py:testFilter`

**StackTrace:**
```
ERROR [o.i.i.i.InvocationContextInterceptor] ISPN000136: Error executing command ReadWriteKeyCommand on Cache 'respCache', writing keys [WrappedByteArray[\d\o\c (3 bytes)]]java.lang.NoClassDefFoundError: net/minidev/json/parser/ParseException
	at com.jayway.jsonpath.internal.filter.ValueNode.createJsonNode(ValueNode.java:207)
	at com.jayway.jsonpath.internal.filter.ValueNodes$PathNode.evaluate(ValueNodes.java:703)
	at com.jayway.jsonpath.internal.filter.RelationalExpressionNode.apply(RelationalExpressionNode.java:37)
	at com.jayway.jsonpath.internal.filter.FilterCompiler$CompiledFilter.apply(FilterCompiler.java:430)
	at com.jayway.jsonpath.internal.path.PredicatePathToken.accept(PredicatePathToken.java:77)
	at com.jayway.jsonpath.internal.path.PredicatePathToken.evaluate(PredicatePathToken.java:60)
	at com.jayway.jsonpath.internal.path.PathToken.handleObjectProperty(PathToken.java:90)
	at com.jayway.jsonpath.internal.path.PropertyPathToken.evaluate(PropertyPathToken.java:80)
	at com.jayway.jsonpath.internal.path.RootPathToken.evaluate(RootPathToken.java:66)
	at com.jayway.jsonpath.internal.path.CompiledPath.evaluate(CompiledPath.java:99)
	at com.jayway.jsonpath.internal.path.CompiledPath.evaluate(CompiledPath.java:107)
	at com.jayway.jsonpath.JsonPath.read(JsonPath.java:191)
	at com.jayway.jsonpath.internal.JsonContext.read(JsonContext.java:88)
	at com.jayway.jsonpath.internal.JsonContext.read(JsonContext.java:77)
	at org.infinispan.server.resp.json.JsonGetFunction.apply(JsonGetFunction.java:91)
	at org.infinispan.server.resp.json.JsonGetFunction.apply(JsonGetFunction.java:27)
	at org.infinispan.interceptors.impl.CallInterceptor.visitReadWriteKeyCommand(CallInterceptor.java:986)
	at org.infinispan.commands.functional.ReadWriteKeyCommand.acceptVisitor(ReadWriteKeyCommand.java:70)
	at org.infinispan.interceptors.impl.CallInterceptor.visitCommand(CallInterceptor.java:151)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:60)
	at org.infinispan.interceptors.BaseAsyncInterceptor.asyncInvokeNext(BaseAsyncInterceptor.java:231)
	at org.infinispan.interceptors.impl.EntryWrappingInterceptor.setSkipRemoteGetsAndInvokeNextForDataCommand(EntryWrappingInterceptor.java:732)
	at org.infinispan.interceptors.impl.EntryWrappingInterceptor.visitReadWriteKeyCommand(EntryWrappingInterceptor.java:512)
	at org.infinispan.commands.functional.ReadWriteKeyCommand.acceptVisitor(ReadWriteKeyCommand.java:70)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextAndFinally(BaseAsyncInterceptor.java:154)
	at org.infinispan.interceptors.locking.AbstractLockingInterceptor.lambda$nonTxLockAndInvokeNext$0(AbstractLockingInterceptor.java:318)
	at org.infinispan.interceptors.SyncInvocationStage.andHandle(SyncInvocationStage.java:71)
	at org.infinispan.interceptors.locking.AbstractLockingInterceptor.nonTxLockAndInvokeNext(AbstractLockingInterceptor.java:313)
	at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitNonTxDataWriteCommand(AbstractLockingInterceptor.java:138)
	at org.infinispan.interceptors.locking.NonTransactionalLockingInterceptor.visitDataWriteCommand(NonTransactionalLockingInterceptor.java:41)
	at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitReadWriteKeyCommand(AbstractLockingInterceptor.java:202)
	at org.infinispan.commands.functional.ReadWriteKeyCommand.acceptVisitor(ReadWriteKeyCommand.java:70)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextThenApply(BaseAsyncInterceptor.java:78)
	at org.infinispan.interceptors.impl.CacheMgmtInterceptor.updateStatisticsReadWrite(CacheMgmtInterceptor.java:332)
	at org.infinispan.interceptors.impl.CacheMgmtInterceptor.visitReadWriteKeyCommand(CacheMgmtInterceptor.java:354)
	at org.infinispan.commands.functional.ReadWriteKeyCommand.acceptVisitor(ReadWriteKeyCommand.java:70)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
	at org.infinispan.interceptors.DDAsyncInterceptor.handleDefault(DDAsyncInterceptor.java:54)
	at org.infinispan.interceptors.DDAsyncInterceptor.visitReadWriteKeyCommand(DDAsyncInterceptor.java:195)
	at org.infinispan.commands.functional.ReadWriteKeyCommand.acceptVisitor(ReadWriteKeyCommand.java:70)
	at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextAndExceptionally(BaseAsyncInterceptor.java:127)
	at org.infinispan.interceptors.impl.InvocationContextInterceptor.visitCommand(InvocationContextInterceptor.java:90)
	at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invokeAsync(AsyncInterceptorChainImpl.java:211)
	at org.infinispan.functional.impl.AbstractFunctionalMap.invokeAsync(AbstractFunctionalMap.java:141)
	at org.infinispan.functional.impl.ReadWriteMapImpl.eval(ReadWriteMapImpl.java:59)
	at org.infinispan.functional.FunctionalMap$ReadWriteMap.eval(FunctionalMap.java:524)
	at org.infinispan.server.resp.json.EmbeddedJsonCache.get(EmbeddedJsonCache.java:63)
	at org.infinispan.server.resp.commands.json.JSONGET.perform(JSONGET.java:40)
	at org.infinispan.server.resp.Resp3Handler.actualHandleRequest(Resp3Handler.java:96)
	at org.infinispan.server.resp.RespRequestHandler.handleRequest(RespRequestHandler.java:88)
	at org.infinispan.server.resp.RespHandler.handleCommandAndArguments(RespHandler.java:194)
	at org.infinispan.server.resp.RespHandler.channelRead(RespHandler.java:170)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at org.infinispan.server.core.transport.StatsChannelHandler.channelRead(StatsChannelHandler.java:28)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:501)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:399)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassNotFoundException: net.minidev.json.parser.ParseException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 76 more
```

### Expected behavior

```
127.0.0.1:6379> JSON.SET doc $ '{"arr": ["kaboom", "kafoosh", "four", "bar", 7.0, "foolish", ["food", "foo", "FoO", "fight"], -9]}'
OK
127.0.0.1:6379> JSON.GET doc '$.arr[?(@ =~ ".*foo")]'
["kafoosh", "foolish"]
127.0.0.1:6379> JSON.SET doc2 $ '{"arr": ["kaboom", "kafoosh", "four", "bar", 7.0, "foolish", -9]}'
OK
127.0.0.1:6379> JSON.GET doc2 '$.arr[?(@ =~ ".*foo")]'
["kafoosh", "foolish"]
```

### Actual behavior

```
127.0.0.1:6379> JSON.SET doc $ '{"arr": ["kaboom", "kafoosh", "four", "bar", 7.0, "foolish", ["food", "foo", "FoO", "fight"], -9]}'
OK
127.0.0.1:6379> JSON.GET doc '$.arr[?(@ =~ ".*foo")]'
(error) ERR net.minidev.json.parser.ParseException
127.0.0.1:6379> JSON.SET doc2 $ '{"arr": ["kaboom", "kafoosh", "four", "bar", 7.0, "foolish", -9]}'
OK
127.0.0.1:6379> JSON.GET doc2 '$.arr[?(@ =~ ".*foo")]'
"[]"
```


### How to Reproduce?

_No response_

--- a/build/configuration/pom.xml
+++ b/build/configuration/pom.xml
@@ -117,6 +117,7 @@
       <version.jgroups>5.5.1.Final</version.jgroups>
       <version.jgroups.raft>1.1.3.Final</version.jgroups.raft>
       <version.json-path>2.10.0</version.json-path>
+      <version.json-smart>2.6.0</version.json-smart>
       <version.jsr107>1.1.0</version.jsr107>
       <version.junit>4.13.2</version.junit>
       <version.junit5>5.14.1</version.junit5>

--- a/documentation/src/main/asciidoc/topics/ref_redis_commands.adoc
+++ b/documentation/src/main/asciidoc/topics/ref_redis_commands.adoc
@@ -159,6 +159,16 @@ NOTE: This implementation attempts to return all attributes that a real Redis se
 * link:https://redis.io/commands/json.forget[JSON.FORGET]
 
 * link:https://redis.io/commands/json.get[JSON.GET]
++
+[NOTE]
+=====
+ JSON filter with RegEx differencies from Redis implementation:
+
+1. Pattern could be expressed in Redis style "(i)pattern" or Java style "/.*pattern.*/i)".
+2. Dynamic filter with regex is not supported. i.e. `$.store.book[?(@.name =~ $.pattern)]` doesn't work.
+3. Filtering an array returns all matching elements and all nested arrays containing matching elements.
+
+=====
 
 * link:https://redis.io/commands/json.mset[JSON.MSET]
 +

--- a/server/resp/pom.xml
+++ b/server/resp/pom.xml
@@ -40,6 +40,11 @@
          <version>${version.json-path}</version>
       </dependency>
       <dependency>
+         <groupId>net.minidev</groupId>
+         <artifactId>json-smart</artifactId>
+         <version>${version.json-smart}</version>
+      </dependency>
+      <dependency>
          <groupId>org.infinispan</groupId>
          <artifactId>infinispan-scripting</artifactId>
       </dependency>

--- a/server/resp/src/main/java/org/infinispan/server/resp/commands/json/JSONCommandArgumentReader.java
+++ b/server/resp/src/main/java/org/infinispan/server/resp/commands/json/JSONCommandArgumentReader.java
@@ -40,7 +40,7 @@ public final class JSONCommandArgumentReader {
     public static CommandArgs readCommandArgs(List<byte[]> arguments,  byte[] key , int pos) {
         byte[] path = arguments.size() > pos ? arguments.get(pos) : DEFAULT_COMMAND_PATH;
         final byte[] jsonPath = JSONUtil.toJsonPath(path);
-        boolean isLegacy = path != jsonPath;
+        boolean isLegacy = !JSONUtil.isJsonPath(path);
         boolean isRoot = path.length == 1 && (path[0] == JSON_ROOT[0] || path[0] == DEFAULT_COMMAND_PATH[0]);
         return new CommandArgs(key, path, jsonPath, isLegacy, isRoot);
     }

--- a/server/resp/src/main/java/org/infinispan/server/resp/json/JSONUtil.java
+++ b/server/resp/src/main/java/org/infinispan/server/resp/json/JSONUtil.java
@@ -1,6 +1,8 @@
 package org.infinispan.server.resp.json;
 
 import java.nio.charset.StandardCharsets;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.jayway.jsonpath.Configuration;
@@ -16,7 +18,8 @@ public class JSONUtil {
    // path.
    // SET operation requires two different config:
    // - definite path needs DEFAULT_PATH_LEAF_TO_NULL otherwise can't add leaf node
-   // - undefinite path doesn't need, because it doesn't add node if missing, it just
+   // - undefinite path doesn't need, because it doesn't add node if missing, it
+   // just
    // update existing node
    public static final Configuration configForDefiniteSet = Configuration.builder().options(Option.SUPPRESS_EXCEPTIONS)
          .options(Option.DEFAULT_PATH_LEAF_TO_NULL)
@@ -33,7 +36,7 @@ public class JSONUtil {
    public static final Configuration configForGet = Configuration.builder().options(Option.ALWAYS_RETURN_LIST)
          .options(Option.SUPPRESS_EXCEPTIONS).jsonProvider(new InfinispanJacksonJsonNodeProvider())
          .mappingProvider(new com.jayway.jsonpath.spi.mapper.JacksonMappingProvider()).build();
-   public static ParseContext parserForGet = JsonPath.using(configForGet);
+   public static final ParseContext parserForGet = JsonPath.using(configForGet);
 
    // Modifier operations need ALWAYS_RETURN_LIST and AS_PATH_LIST
    public static final Configuration configForMod = Configuration.builder().options(Option.AS_PATH_LIST)
@@ -56,33 +59,65 @@ public class JSONUtil {
    }
 
    /**
-    * Returns a jsonpath version of path
+    * Returns a jsonpath version of path with Redis regex patterns converted to
+    * Jayway format.
     *
-    * @return path itself if path is already in Jsonpath format or the jsonpath equivalent in a new
-    *         byte[]
+    * This method performs two conversions in order: 1. Legacy path format
+    * conversion (e.g., ".foo"
+    * → "$.foo") 2. Redis regex format conversion on the result (e.g., =~ "pattern"
+    * → =~ /pattern/)
+    *
+    * @param path
+    *             the path as a String
+    * @return path converted to Jayway-compatible JSONPath format
+    */
+   public static String toJsonPath(String path) {
+      // Step 1: Handle legacy path format conversion (. → $)
+      String step1Result;
+      if (!isJsonPath(path)) {
+         if (path.startsWith(".")) {
+            if (path.length() == 1) {
+               // For '.' return json root '$'
+               step1Result = "$";
+            } else {
+               // Just prepend $ to the beginning
+               step1Result = "$" + path;
+            }
+         } else {
+            // prepend $. to the beginning
+            // Using "$." so wrong legacy path like "" and " " will fail
+            step1Result = "$." + path;
+         }
+      } else {
+         step1Result = path;
+      }
+
+      // Step 2: Convert Redis regex format to Jayway format
+      return convertRedisFilterToJayway(step1Result);
+   }
+
+   /**
+    * Returns a jsonpath version of path with Redis regex patterns converted to
+    * Jayway format.
+    *
+    * This is a convenience method that converts byte[] to String, performs the
+    * conversions, and
+    * returns the result as byte[].
+    *
+    * @param path
+    *             the path as byte[]
+    * @return path converted to Jayway-compatible JSONPath format as byte[]
     */
    public static byte[] toJsonPath(byte[] path) {
-      if (!isJsonPath(path)) {
-         if (path.length >= 1 && path[0] == '.') {
-            if (path.length == 1) {
-               // For '.' return json root '$'
-               return JSON_ROOT;
-            }
-            // Just append $ to the beginning
-            byte[] result = new byte[(path.length + 1)];
-            result[0] = JSON_ROOT_BYTE;
-            System.arraycopy(path, 0, result, 1, path.length);
-            return result;
-         }
-         // append $. to the beginning
-         // Using "$." so wrong legacy path like "" and " " will fail
-         byte[] result = new byte[(path.length + 2)];
-         result[0] = JSON_ROOT_BYTE;
-         result[1] = '.';
-         System.arraycopy(path, 0, result, 2, path.length);
-         return result;
+      // Special case optimization: single '.' becomes '$' without String conversion
+      if (path != null && path.length == 1 && path[0] == '.') {
+         return JSON_ROOT;
       }
-      return path;
+
+      // Convert to String, apply conversions, convert back
+      String pathStr = new String(path, StandardCharsets.UTF_8);
+      String converted = toJsonPath(pathStr);
+      return converted.equals(pathStr) ? path : converted.getBytes(StandardCharsets.UTF_8);
    }
 
    public static boolean isJsonPath(byte[] path) {
@@ -96,44 +131,109 @@ public class JSONUtil {
 
    // Invalid values for Redis. Expecially '\0xa' breaks RESP, seen as end of data
    public static boolean isValueInvalid(byte[] value) {
-      if (value.length == 0) return true;
+      if (value.length == 0)
+         return true;
       if (value.length == 1) {
-          return isSingleCharInvalid(value[0]);
+         return isSingleCharInvalid(value[0]);
       }
       if (value.length == 2) {
-          return isDoubleCharInvalid(value);
+         return isDoubleCharInvalid(value);
       }
       return false;
-  }
+   }
 
-  private static boolean isSingleCharInvalid(byte value) {
+   private static boolean isSingleCharInvalid(byte value) {
       switch (value) {
-          case ' ':
-          case '{':
-          case '}':
-          case '[':
-          case ']':
-          case '\\':
-          case '\'':
-          case 0:
-          case 0x0a:
-          case 0x0c:
-              return true;
-          default:
-              return false;
+         case ' ':
+         case '{':
+         case '}':
+         case '[':
+         case ']':
+         case '\\':
+         case '\'':
+         case 0:
+         case 0x0a:
+         case 0x0c:
+            return true;
+         default:
+            return false;
       }
-  }
+   }
 
-  private static boolean isDoubleCharInvalid(byte[] value) {
+   private static boolean isDoubleCharInvalid(byte[] value) {
       if (value[0] == '\\' && (value[1] == '\\' || value[1] == '"' || value[1] == '[')) {
-          return true;
+         return true;
       }
       if (value[0] == '{' && value[1] == ']') {
-          return true;
+         return true;
       }
       if (value[0] == '[' && value[1] == '}') {
-          return true;
+         return true;
       }
       return false;
-  }
+   }
+
+   // --- Base regex converter (from before) ---
+   private static String toJaywayRegex(String redisRegex) {
+      if (redisRegex == null || redisRegex.isBlank()) {
+         throw new IllegalArgumentException("Regex cannot be null or empty");
+      }
+
+      // Trim quotes if present
+      String pattern = redisRegex.trim();
+      boolean isQuoted = pattern.startsWith("\"") && pattern.endsWith("\"");
+      if (isQuoted && pattern.length() >= 2) {
+         pattern = pattern.substring(1, pattern.length() - 1);
+      }
+
+      boolean startsWithAnchor = pattern.startsWith("^");
+      boolean endsWithAnchor = pattern.endsWith("$");
+      boolean hasLeadingWildcard = pattern.startsWith(".*");
+      boolean hasTrailingWildcard = pattern.endsWith(".*");
+
+      if (isQuoted) {
+         // Add missing wildcards only if needed
+         if (!startsWithAnchor && !hasLeadingWildcard) {
+            pattern = ".*" + pattern;
+         }
+         if (!endsWithAnchor && !hasTrailingWildcard) {
+            pattern = pattern + ".*";
+         }
+         // Wrap in delimiters
+         return "/" + pattern + "/";
+      }
+      return pattern;
+   }
+
+   // --- Filter-level converter ---
+   /**
+    * Converts any RedisJSON-style =~ "regex" fragments into Jayway =~ /regex/
+    * format. Keeps the
+    * rest of the JSONPath filter untouched.
+    *
+    * Example: $[?(@.name =~ ".*foo" && @.type =~ "(?i)^bar$")] to $[?(@.name =~
+    * /.*foo.*\/
+    * && @.type=~/(?i)^bar$/)]
+    */
+
+   public static String convertRedisFilterToJayway(String redisFilter) {
+      if (redisFilter == null || redisFilter.isBlank()) {
+         throw new IllegalArgumentException("Filter cannot be null or empty");
+      }
+
+      // Match patterns like =~ "something"
+      Pattern p = Pattern.compile("=~\\s*\"([^\"]+)\"");
+      Matcher m = p.matcher(redisFilter);
+
+      StringBuilder sb = new StringBuilder();
+      while (m.find()) {
+         String redisRegex = "\"" + m.group(1) + "\"";
+         String jaywayRegex = toJaywayRegex(redisRegex);
+         m.appendReplacement(sb, "=~ " + jaywayRegex);
+      }
+      m.appendTail(sb);
+
+      return sb.toString();
+   }
+
 }

--- a/server/resp/src/main/java/org/infinispan/server/resp/json/JsonGetFunction.java
+++ b/server/resp/src/main/java/org/infinispan/server/resp/json/JsonGetFunction.java
@@ -81,7 +81,7 @@ public class JsonGetFunction
          // Convert to jsonpath and set legacy=false if any path is a jsonpath
          List<byte[]> jsonPaths = paths.stream().map((p) -> {
             var jp = JSONUtil.toJsonPath(p);
-            isLegacy &= (jp != p);
+            isLegacy = isLegacy && !JSONUtil.isJsonPath(p);
             return jp;
          }).toList();
 

--- a/server/resp/src/test/java/org/infinispan/server/resp/JsonFilterCommandsTest.java
+++ b/server/resp/src/test/java/org/infinispan/server/resp/JsonFilterCommandsTest.java
@@ -1,0 +1,233 @@
+package org.infinispan.server.resp;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.infinispan.test.TestingUtil.k;
+
+import java.util.List;
+
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import io.lettuce.core.api.sync.RedisCommands;
+import io.lettuce.core.json.DefaultJsonParser;
+import io.lettuce.core.json.JsonPath;
+import io.lettuce.core.json.JsonValue;
+
+@Test(groups = "functional", testName = "server.resp.JsonFilterCommandsTest")
+public class JsonFilterCommandsTest extends SingleNodeRespBaseTest {
+   /**
+    * RESP JSON filter commands testing
+    *
+    * @since 15.2
+    */
+   private RedisCommands<String, String> redis;
+   private DefaultJsonParser defaultJsonParser = new DefaultJsonParser();
+
+   @BeforeMethod
+   public void initConnection() {
+      redis = redisConnection.sync();
+   }
+
+   @Test
+   public void testJSONGETWithFilterExpression() {
+      // Test JSONPath filter expressions with regex matching
+      JsonPath jpRoot = new JsonPath("$");
+      String key = k();
+
+      // JSON.SET doc $ '{"arr":["foo","bar","foobar","bazfoo","baz"]}'
+      JsonValue jv = defaultJsonParser.createJsonValue("""
+            {"arr":["foo","bar","foobar","bazfoo","baz"]}
+            """);
+      assertThat(redis.jsonSet(key, jpRoot, jv)).isEqualTo("OK");
+
+      // JSON.GET doc '$.arr[?(@ =~ ".*foo")]'
+      // Should return elements matching the regex pattern ".*foo"
+      JsonPath filterPath = new JsonPath("$.arr[?(@ =~ \".*foo\")]");
+      List<JsonValue> result = redis.jsonGet(key, filterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(3);
+      assertThat(result.get(0).asJsonArray().asList())
+            .map(JsonValue::asString)
+            .containsExactly("foo", "foobar", "bazfoo");
+
+      // Same test with Java regex syntax
+      JsonPath javaFilterPath = new JsonPath("$.arr[?(@ =~ /.*foo.*/)]");
+      result = redis.jsonGet(key, javaFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(3);
+      assertThat(result.get(0).asJsonArray().asList())
+            .map(JsonValue::asString)
+            .containsExactly("foo", "foobar", "bazfoo");
+
+      // Test with different filter: exact match
+      // JSON.GET doc $.arr[?(@ == "bar")]
+      JsonPath exactMatchPath = new JsonPath("$.arr[?(@ == \"bar\")]");
+      result = redis.jsonGet(key, exactMatchPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(1);
+      assertThat(result.get(0).asJsonArray().getFirst().asString()).isEqualTo("bar");
+
+      // Test with numeric array and comparison filter
+      // JSON.SET doc2 $ '{"nums":[1,5,10,15,20,25]}'
+      String key2 = k(2);
+      JsonValue jv2 = defaultJsonParser.createJsonValue("""
+            {"nums":[1,5,10,15,20,25]}
+            """);
+      assertThat(redis.jsonSet(key2, jpRoot, jv2)).isEqualTo("OK");
+
+      // JSON.GET doc2 $.nums[?(@ > 10)]
+      JsonPath numFilterPath = new JsonPath("$.nums[?(@ > 10)]");
+      result = redis.jsonGet(key2, numFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(3);
+      assertThat(result.get(0).asJsonArray().asList())
+            .map(JsonValue::asNumber)
+            .map(Number::intValue)
+            .containsExactly(15, 20, 25);
+
+      // Test with object array and nested property filter
+      // JSON.SET doc3 $ '{"items":[{"name":"foo","price":10},{"name":"bar","price":20},{"name":"foobar","price":15}]}'
+      String key3 = k(3);
+      JsonValue jv3 = defaultJsonParser.createJsonValue("""
+            {"items":[{"name":"foo","price":10},{"name":"bar","price":20},{"name":"foobar","price":15}]}
+            """);
+      assertThat(redis.jsonSet(key3, jpRoot, jv3)).isEqualTo("OK");
+
+      // JSON.GET doc3 $.items[?(@.price > 10)]
+      JsonPath objFilterPath = new JsonPath("$.items[?(@.price > 10)]");
+      result = redis.jsonGet(key3, objFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(2);
+      // Should return the two items with price > 10
+      String resultStr = result.get(0).toString();
+      assertThat(resultStr).isEqualTo("[{\"name\":\"bar\",\"price\":20},{\"name\":\"foobar\",\"price\":15}]");
+
+      // Test with complex array and regex filter
+      // JSON.SET doc4 $ '{"arr": ["kaboom", "kafoosh", "four", "bar", 7.0, "foolish", ["FoO", "fight"]]}'
+      String key4 = k(4);
+      JsonValue jv4 = defaultJsonParser.createJsonValue("""
+            {"arr": ["kaboom", "kafoosh", "four", "bar", 7.0, "foolish", ["FoO", "fight"]]}
+            """);
+      assertThat(redis.jsonSet(key4, jpRoot, jv4)).isEqualTo("OK");
+
+      // JSON.GET doc4 '$.arr[?(@ =~ "foo")]' - should match "kafoosh", "foolish" (case-sensitive partial match)
+      JsonPath fooFilterPath = new JsonPath("$.arr[?(@ =~ \"foo\")]");
+      result = redis.jsonGet(key4, fooFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(2);
+      assertThat(result.get(0).asJsonArray().asList())
+            .map(JsonValue::asString)
+            .containsExactly("kafoosh", "foolish");
+
+      // JSON.GET doc4 '$.arr[?(@ =~ "^ka")]' - should match "kaboom", "kafoosh" (starts with "ka")
+      JsonPath kaFilterPath = new JsonPath("$.arr[?(@ =~ \"^ka\")]");
+      result = redis.jsonGet(key4, kaFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(2);
+      assertThat(result.get(0).asJsonArray().asList())
+            .map(JsonValue::asString)
+            .containsExactly("kaboom", "kafoosh");
+   }
+
+   @Test
+   public void testJSONGETWithDynamicFilter() {
+      // Test JSONPath filter expressions that compare document fields
+      JsonPath jpRoot = new JsonPath("$");
+      String key = k();
+
+      // Test comparing two numeric fields in the same object
+      // JSON.SET doc $ '{"products":[{"name":"laptop","price":1000,"discount":100},{"name":"mouse","price":50,"discount":60},{"name":"keyboard","price":80,"discount":20}]}'
+      JsonValue jv = defaultJsonParser.createJsonValue("""
+            {"products":[
+               {"name":"laptop","price":1000,"discount":100},
+               {"name":"mouse","price":50,"discount":60},
+               {"name":"keyboard","price":80,"discount":20}
+            ]}
+            """);
+      assertThat(redis.jsonSet(key, jpRoot, jv)).isEqualTo("OK");
+
+      // JSON.GET doc '$.products[?(@.price > @.discount)]' - should return items where price > discount
+      JsonPath filterPath = new JsonPath("$.products[?(@.price > @.discount)]");
+      List<JsonValue> result = redis.jsonGet(key, filterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(2);
+      // Should return laptop and keyboard, but not mouse (50 < 60)
+      String resultStr = result.get(0).toString();
+      assertThat(resultStr).contains("laptop");
+      assertThat(resultStr).contains("keyboard");
+      assertThat(resultStr).doesNotContain("mouse");
+
+      // Test with equality comparison
+      // JSON.SET doc2 $ '{"users":[{"name":"alice","age":30,"minAge":25},{"name":"bob","age":20,"minAge":20},{"name":"charlie","age":35,"minAge":40}]}'
+      String key2 = k(2);
+      JsonValue jv2 = defaultJsonParser.createJsonValue("""
+            {"users":[
+               {"name":"alice","age":30,"minAge":25},
+               {"name":"bob","age":20,"minAge":20},
+               {"name":"charlie","age":35,"minAge":40}
+            ]}
+            """);
+      assertThat(redis.jsonSet(key2, jpRoot, jv2)).isEqualTo("OK");
+
+      // JSON.GET doc2 '$.users[?(@.age >= @.minAge)]' - should return users where age >= minAge
+      JsonPath ageFilterPath = new JsonPath("$.users[?(@.age >= @.minAge)]");
+      result = redis.jsonGet(key2, ageFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(2);
+      resultStr = result.get(0).toString();
+      assertThat(resultStr).contains("alice");
+      assertThat(resultStr).contains("bob");
+      assertThat(resultStr).doesNotContain("charlie");
+
+      // Test with string field comparison
+      // JSON.SET doc3 $ '{"items":[{"id":"A","category":"electronics","targetCategory":"electronics"},{"id":"B","category":"books","targetCategory":"electronics"},{"id":"C","category":"toys","targetCategory":"toys"}]}'
+      String key3 = k(3);
+      JsonValue jv3 = defaultJsonParser.createJsonValue("""
+            {"items":[
+               {"id":"A","category":"electronics","targetCategory":"electronics"},
+               {"id":"B","category":"books","targetCategory":"electronics"},
+               {"id":"C","category":"toys","targetCategory":"toys"}
+            ]}
+            """);
+      assertThat(redis.jsonSet(key3, jpRoot, jv3)).isEqualTo("OK");
+
+      // JSON.GET doc3 '$.items[?(@.category == @.targetCategory)]' - should return items where category matches targetCategory
+      JsonPath categoryFilterPath = new JsonPath("$.items[?(@.category == @.targetCategory)]");
+      result = redis.jsonGet(key3, categoryFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(2);
+      resultStr = result.get(0).toString();
+      assertThat(resultStr).contains("\"id\":\"A\"");
+      assertThat(resultStr).contains("\"id\":\"C\"");
+      assertThat(resultStr).doesNotContain("\"id\":\"B\"");
+
+      // Test with complex comparison combining field comparison and constant
+      // JSON.SET doc4 $ '{"scores":[{"player":"alice","score":100,"threshold":80,"bonus":10},{"player":"bob","score":70,"threshold":80,"bonus":5},{"player":"charlie","score":90,"threshold":85,"bonus":15}]}'
+      String key4 = k(4);
+      JsonValue jv4 = defaultJsonParser.createJsonValue("""
+            {"scores":[
+               {"player":"alice","score":100,"threshold":80,"bonus":10},
+               {"player":"bob","score":70,"threshold":80,"bonus":5},
+               {"player":"charlie","score":90,"threshold":85,"bonus":15}
+            ],
+            "pattern": "bob"}
+            """);
+      assertThat(redis.jsonSet(key4, jpRoot, jv4)).isEqualTo("OK");
+
+      // JSON.GET doc4 '$.scores[?(@.score > @.threshold && @.bonus > 10)]' - combined filter
+      JsonPath combinedFilterPath = new JsonPath("$.scores[?(@.score > @.threshold && @.bonus > 10)]");
+      result = redis.jsonGet(key4, combinedFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(1);
+      resultStr = result.get(0).toString();
+      assertThat(resultStr).contains("charlie");
+      assertThat(resultStr).doesNotContain("alice");
+      assertThat(resultStr).doesNotContain("bob");
+
+      // Test with pattern not in the current node
+      JsonPath dynFilterPath = new JsonPath("$.scores[?(@.player == $.pattern)]");
+      result = redis.jsonGet(key4, dynFilterPath);
+      assertThat(result).hasSize(1);
+      assertThat(result.get(0).asJsonArray().size()).isEqualTo(1);
+   }
+}

--- a/server/tests/src/test/java/org/infinispan/server/footprint/FootprintIT.java
+++ b/server/tests/src/test/java/org/infinispan/server/footprint/FootprintIT.java
@@ -31,8 +31,8 @@ public class FootprintIT {
    private static final int LOADED_CLASS_COUNT_UPPER_BOUND = 11_900;
    private static final long HEAP_USAGE_LOWER_BOUND = 24_500_000L;
    private static final long HEAP_USAGE_UPPER_BOUND = 28_000_000L;
-   private static final long DISK_USAGE_LOWER_BOUND = 76_000_000L;
-   private static final long DISK_USAGE_UPPER_BOUND = 77_500_000L;
+   private static final long DISK_USAGE_LOWER_BOUND = 77_500_000L;
+   private static final long DISK_USAGE_UPPER_BOUND = 79_000_000L;
 
    private static final int JACOCO_CLASS_COUNT = 165;
    private static final long JACOCO_HEAP_USAGE = 2_350_000L;

